File: 01 Rational Numbers/Rational.java
// Steven Abreu
// Jacobs University Bremen
// AaDS Fall 2016
// Problem 1.2

import java.math.BigInteger;


class RationalNumber {

  private BigInteger d, e;

  RationalNumber(BigInteger d, BigInteger e) {
    this.d = d;
    // take the absolute value
    if (e.equals(new BigInteger("0"))) {
      System.out.println("WARNING: denominator cannot be zero -> set to 1.");
      this.e = new BigInteger("1");
    } else {
      this.e = e.abs();
    }
  }

  BigInteger getD(){ return d; }

  BigInteger getE(){ return e; }

  void setD(BigInteger i) { d = i; }

  void setE(BigInteger i) { e = i.abs(); }

  Boolean equal(RationalNumber r) {
    if (d.multiply(r.getE()) == e.multiply(r.getD())) {
      return true;
    } else {
      return false;
    }
  }

  RationalNumber getNormal() {
    BigInteger g = gcd(d,e);
    return new RationalNumber(d.divide(g), e.divide(g));
  }

  void makeNormal() {
    BigInteger g = gcd(d,e);
    d = d.divide(g);
    e = e.divide(g);
  }

  // STATIC METHODS
  static BigInteger zero = new BigInteger("0");

  static BigInteger gcd(BigInteger a, BigInteger b) {
    BigInteger r = a.mod(b);
    while (!r.equals(zero)) {
      a = b;
      b = r;
      r = a.mod(b);
    }
    return b;
  }
}

public class Rational {

  public RationalNumber addition(RationalNumber r, RationalNumber s) {
    BigInteger d = r.getD().multiply(s.getE()).add(r.getE().multiply(s.getD()));
    BigInteger e = r.getE().multiply(s.getE());
    return new RationalNumber(d,e);
  }

  public RationalNumber additiveInverse(RationalNumber r) {
    return new RationalNumber(r.getD().multiply(new BigInteger("-1")),r.getE());
  }

  public RationalNumber multiply(RationalNumber r, RationalNumber s) {
    return new RationalNumber(r.getD().multiply(s.getE()),r.getE().multiply(s.getE()));
  }

  public RationalNumber multiplicativeInverse(RationalNumber r) {
    return new RationalNumber(r.getE(), r.getD());
  }

  public static void main(String[] args) {
    BigInteger a = new BigInteger("2");
    BigInteger b = new BigInteger("0");
    new RationalNumber(a,b);
  }
}

File: Algorithms Data Structures/01 Rational Numbers/Rational.py
# Steven Abreu
# Jacobs University Bremen
# AaDS Fall 2016
# Problem 1.2

def gcd(a,b):
    r = a % b
    while r != 0:
        a = b
        b = r
        r = a % b
    return b

class Rational:
    def __init__(self,d,e):
        self.d = d
        # take the absolute value
        if e == 0:
            # if input is zero -> do 1 instead
            print("WARNING: denominator cannot be zero -> set to 1.")
            self.e = 1
        else:
            self.e = abs(e)
    def getD(self):
        return self.d
    def getE(self):
        return self.e
    def equal(self,r):
        if self.d * r.getE() == self.e * r.getD():
            return True
        else:
            return False
    def getNormal(self):
        g = gcd(self.d,self.e)
        d1 = self.d/g
        e1 = self.e/g
        return Rational(d1,e1)
    def makeNormal(self):
        g = gcd(self.d,self.e)
        self.d = self.d/g
        self.e = self.e/g
    def __str__(self):
        return str(self.d) + "/" + str(self.e)

def addition(r,s):
    d = r.getD() * s.getE() + r.getE() * s.getD()
    e = r.getE() * s.getE()
    return Rational(d,e)

def additiveInverse(r):
    return Rational(-1*r.getD(),r.getE())

def multiplication(r,s):
    return Rational(r.getD() * s.getD(), r.getE() * s.getE())

def multiplicativeInverse(r):
    return Rational(r.getE(), r.getD())
